src/GafferSceneUI/Inspector.cpp

```diff
@@ -589,6 +592,135 @@ PathPtr Inspector::historyPath()
+
+	size_t historySize( const IECore::Canceller *canceller )
+	{
+		std::call_once( m_initFlag, &HistoryProvider::initHistory, this, canceller );
+		return m_historyVector.size();
+	}
+
+	const SceneAlgo::History *historyItem( size_t index, const IECore::Canceller *canceller  )
+	{
+		std::call_once( m_initFlag, &HistoryProvider::initHistory, this, canceller );
+		return m_historyVector[index].get();
+	}
+
+	private :
+
+		const ConstContextPtr m_context;
+		std::thread::id m_constructorThreadId;
+
+		void initHistory( const IECore::Canceller *canceller )
+		{
+			assert( m_historyVector.empty() );
+
+			if( std::this_thread::get_id() == m_constructorThreadId )
+			{
+				// Our intention is that HistoryPath is constructed on the main
+				// thread and then only ever evaluated in BackgroundTasks by a
+				// PathListingWidget, to avoid blocking the UI. Warn if we fail
+				// in this regard.
+				IECore::msg( IECore::Msg::Warning, "HistoryPath", "Path evaluated on unexpected thread" );
+			}
+
+			Context::EditableScope scope( m_context.get() );
+			if( canceller )
+			{
+				scope.setCanceller( canceller );
+			}
+
+			SceneAlgo::History::ConstPtr history = inspector->history();
+			if( !history )
+			{
+				return;
+			}
+
+			std::string editWarning;
+
+			ConstObjectPtr successorValue;
+			const SceneAlgo::History *successor = nullptr;
+			const SceneAlgo::History *current = history.get();
+			while( current )
+			{
+				Context::EditableScope scope( current->context.get() );
+				if( canceller )
+				{
+					scope.setCanceller( canceller );
+				}
+
+				ConstObjectPtr currentValue = inspector->value( current );
+				if( successor && !endsWith( m_historyVector, successor ) )
+				{
+					// The result of `inspector->source()` was null on the
+					// previous (`successor`) iteration. But there may still
+					// have been a value change at that point, in which case the
+					// `successor` belongs in our history.
+					if(
+						(bool)currentValue != (bool)successorValue ||
+						( currentValue && currentValue->isNotEqualTo( successorValue.get() ) )
+					)
+					{
+						m_historyVector.push_back( successor );
+					}
+				}
+
+				if( inspector->source( current, editWarning ) )
+				{
+					m_historyVector.push_back( current );
+				}
+
+				if( current->predecessors.size() > 1 )
+				{
+					IECore::msg(
+						IECore::Msg::Warning,
+						"Inspector::HistoryPath",
+						"Branching histories are not supported, using first predecessor history only."
+					);
+				}
+
+				successor = current;
+				successorValue = currentValue;
+				current = current->predecessors.size() ? current->predecessors[0].get() : nullptr;
+				if( !current && !endsWith( m_historyVector, successor ) )
+				{
+					// Make sure we include the tip of the history whether or not
+					// there is an edit there. Among other things this allows us to
+					// show where a value is loaded from a SceneReader.
+					m_historyVector.push_back( successor );
+				}
```

Comment:
This is a great improvement though I find this to be a little confusing when inspecting the history of a fallback value. The history window will display the SceneReader or first node to create that location with the fallback value alongside, but if you then create an edit the SceneReader/location source disappears and the history only displays that first node to actually author a value...

Not something that needs to be addressed immediately, but may end up on the list of further improvements to presenting the history, such as how do we display history for inherited values...

Reply:
I haven't checked, but I _think_ the culprit here is `SceneAlgo::attributeHistory()` rather than AttributeInspector or HistoryPath. Definitely worth a look - I'll put it on my list.